src/semantics/context.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Semantic Analysis Context
+ *
+ * A context object holds state for the semantic analysis phase, such as the
+ * enclosing function (if any), whether or not we are in a loop, a map of
+ * declarations introduced in this scope, and the parent context.
+ *
+ *   import Context from './semantics/context';
+ */
+
+import FunctionDeclaration from '../ast/function-declaration';
+import FunctionObject from '../ast/function-object';
+import Parameter from '../ast/parameter';
+import { Referent } from '../type-definitions/plainscript-types';
+
+interface IContextSchema {
+    parent?: Context | null;
```

Comment:
Why `?` _and_ `| null`? Can this be simplified?

Reply:
Good question! Nullable types are implemented with unions ! The optional parameter operator `?` only adds an undefined type to the parameter, not a null.

![image](https://user-images.githubusercontent.com/16827894/73142368-82851380-4042-11ea-9fe6-b9e7435651b0.png)

Reply:
from [TypeScript's Advanced Types docs](https://www.typescriptlang.org/docs/handbook/advanced-types.html)

Reply:
Thank you, I understand now. Makes perfect sense!